# PS3 Bonus Question

## 1. How many Spam Emails are there?

In [39]:
def count_spam(path = "D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data", folders = ['enron1','enron2','enron3']):
    import os
    from pathlib import Path
    #path = Path("D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data")
    path = Path(path)
    spam_count = 0
    for folder in folders:
        file_names = os.listdir(path / folder / 'spam')
        for file_name in file_names:
            if 'spam.txt' in file_name:
                spam_count += 1
    print('There are', spam_count, 'spam emails in', folders)

In [40]:
count_spam()

There are 4492 spam emails in ['enron1', 'enron2', 'enron3']


## 2. Structure the email data from the 2 directories into 1 dataframe with columns: Status, Subject, Body.

#### Specify the folders that will be included and the path leading to the folders.

In [3]:
folders = ['enron1','enron2','enron3']
status = ['spam','ham']
path1 = "D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\{folder}\\{status}"

#### Create the paths containing all the desired emails.

In [4]:
from itertools import product
directories = list(product(folders, status))
directories

[('enron1', 'spam'),
 ('enron1', 'ham'),
 ('enron2', 'spam'),
 ('enron2', 'ham'),
 ('enron3', 'spam'),
 ('enron3', 'ham')]

In [5]:
paths = []
for i in directories:
    paths.append(path1.format(folder=i[0],status=i[1]))
paths

['D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\spam',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\ham',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron2\\spam',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron2\\ham',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron3\\spam',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron3\\ham']

#### List all the files in each `path`.

In [6]:
import glob
file_paths = [glob.glob(pathname=path + '\\*') for path in paths]

#### Now we have all the file paths of the emails in the specified folders.

In [7]:
file_paths[0][0:5]

['D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\spam\\0006.2003-12-18.GP.spam.txt',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\spam\\0008.2003-12-18.GP.spam.txt',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\spam\\0017.2003-12-18.GP.spam.txt',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\spam\\0018.2003-12-18.GP.spam.txt',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\spam\\0026.2003-12-18.GP.spam.txt']

In [8]:
file_paths[1][0:5]

['D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\ham\\0001.1999-12-10.farmer.ham.txt',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\ham\\0002.1999-12-13.farmer.ham.txt',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\ham\\0003.1999-12-14.farmer.ham.txt',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\ham\\0004.1999-12-14.farmer.ham.txt',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\ham\\0005.1999-12-14.farmer.ham.txt']

#### Having acquired the file paths, we can now extract the status, subject and message body of the emails.

In [9]:
subjects = []
bodies = []
status = []

# for each path
for file_path in file_paths:
    # for each file in path
    for i in file_path:
        # for status
        if 'spam.txt' in i:
            status.append('spam')
        elif 'ham.txt' in i:
            status.append('ham')
        # extract the subject from email
        with open(i,'r',errors='ignore') as f:
            subject = f.readlines()[0]
            subject = subject[9:]
            subjects.append(subject)
        # extract body message from email
        with open(i,'r',errors='ignore') as f:
            body = f.readlines()
            body = body[1:]
            bodies.append(body)
            
# join each sub-elements of the bodies list
bodies = [' '.join(i) for i in bodies]

print('Status has', len(status), 'elements.')
print('Subject has', len(subjects), 'elements.')
print('Body has', len(bodies), 'elements.')

Status has 16537 elements.
Subject has 16537 elements.
Body has 16537 elements.


#### Now, use the lists to create a dataframe.

In [10]:
import pandas as pd

In [11]:
enron = pd.DataFrame(list(zip(status,subjects,bodies)), columns=['Status','Subject','Body'])
enron.sample(10)

Status                                           Subject  \
2787     ham              re : coastal oil & gas corporation\n   
6432    spam  ms 2003 software titles available for download\n   
11559   spam             harvard tech - great computer deals\n   
15502    ham                              enrononline budget\n   
5210    spam           traditional & internet marketing tool\n   
8971     ham              year end 2000 performance feedback\n   
12749    ham              re : what happened to the letter ?\n   
15433    ham                                            fw :\n   
15112    ham     re : gtc change for removal of hpl language\n   
6415    spam                                       vaal medz\n   

                                                    Body  
2787   daren ,\n fyi .\n bob\n - - - - - - - - - - - ...  
6432   opt - in email special offer unsubscribe me se...  
11559  why make them richer\n we have all your favori...  
15502  louise - -\n i would like to meet with you in ...  
5210   the ultimate traditional & internet marketing ...  
8971   note : you will receive this message each time...  
12749  louise ,\n greg signed it and it went out tues...  
15433  - - - - - original message - - - - -\n from : ...  
15112  hi mark , barbara ,\n can we work something ou...  
6415   how t pestilent o save on your medlcations ove...

## 3. Build a Naive Bayes classifier to classify whether emails are spam or not.

#### Import libraries

In [12]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

In [13]:
enron.sample(5)

Status                                 Subject  \
12755    ham                        kern expansion\n   
13351    ham          california update 6 / 5 / 01\n   
652     spam  hiv , hiv , charset = us - ascii " >\n   
2766     ham          calpine daily gas nomination\n   
11078   spam                           re [ 19 ] :\n   

                                                    Body  
12755  fyi folks -\n the kern river expansion that we...  
13351  executive summary\n ? edison ' s mou condemned...  
652    save up to 80 % on popular meds !\n * * * grea...  
2766   >\n ricky a . archer\n fuel supply\n 700 louis...  
11078  msn . com let me do it online lockerbie blamec...

In [14]:
enron.shape

(16537, 3)

#### Create `Spam` column from the `Status` column and delete `Status` column.

In [15]:
enron['Spam'] = [1 if i == 'spam' else 0 for i in enron['Status']]
enron.drop('Status', inplace=True, axis=1)

In [16]:
enron.sample(5)

Subject  \
2330                  ces volumes on tejas\n   
14735  western power update 10 / 10 / 2001\n   
9548               re : biliana ' s resume\n   
1471                    thank god hes gone\n   
16309                 peoples energy / ubs\n   

                                                    Body  Spam  
2330                june volumes on tejas are attached .     0  
14735  for questions or comments regarding this repor...     0  
9548   biliana ,\n i am glad i could help .\n look fo...     0  
1471   come and checkout my new site ! i ' ve used so...     1  
16309  in a strategic move to provide relative probab...     0

#### Split data for training and testing.

In [17]:
from sklearn.model_selection import train_test_split

In [18]:
train_X, test_X, train_y, test_y = train_test_split(enron["Body"], enron["Spam"], test_size=0.3, random_state=0)

#### Import libraries for text analysis

In [19]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import GaussianNB

In [20]:
vectorizer = TfidfVectorizer(stop_words='english')
tfidf_train_X = vectorizer.fit_transform(train_X)

In [21]:
index_value={i[1]:i[0] for i in vectorizer.vocabulary_.items()}

In [22]:
fully_indexed = []
for row in tfidf_train_X:
    fully_indexed.append({index_value[column]:value for (column,value) in zip(row.indices,row.data)})

#### Train the model

In [23]:
classifier = GaussianNB()
classifier.fit(tfidf_train_X.toarray(), train_y)

GaussianNB()

## 4. What is the longest ham email?

#### Create a dataframe containing all ham emails.

In [24]:
hams = enron[enron['Spam'] == 0][['Body']]

#### Create a column containing the lengths of each emails.

In [25]:
hams['Lengths'] = [len(i) for i in hams['Body']]
hams.drop('Body',inplace=True,axis=1)

#### Get the index of the longest ham email.

In [26]:
longest_ham = hams.idxmax()
print(longest_ham)

Lengths    15750
dtype: int64


#### Print the longest ham email.

In [27]:
enron.iloc[longest_ham,:].values

array([['enron mentions\n',
        'enron : a wake - up call\n the wall street journal , 12 / 04 / 01\n fall of a power giant : dynegy hits back at enron with lawsuit - - - claim is that failed merger entitles concern to get northern natural gas\n the wall street journal , 12 / 04 / 01\n enron secures $ 1 . 5 billion in new financing - - - j . p . morgan , citigroup pact is first step ; now , lifeline is sought for trading unit\n the wall street journal , 12 / 04 / 01\n fall of a power giant : burst of chapter 11 filings marks big year for the bankruptcy bar\n the wall street journal , 12 / 04 / 01\n fall of a power giant : now is the time for creditors to vie against each other\n the wall street journal , 12 / 04 / 01\n auditor andersen is added to enron 401 ( k ) lawsuit\n the wall street journal , 12 / 04 / 01\n stadium jinx : what to call enron field ? ` enron folds , \' maybe - - - firms that put their names on arenas hit hard times ; former trans world dome\n the wall street jou

#### The longest ham email has `43888` characters.

In [28]:
hams.max()

Lengths    230120
dtype: int64

## 5. What is the accuracy of your model?

In [29]:
from sklearn.model_selection import cross_val_score

In [30]:
tfidf_test_X = vectorizer.transform(test_X)
print(tfidf_test_X.shape)
scores = cross_val_score(classifier, tfidf_test_X.toarray(), test_y, cv=5)
acc = scores.mean()
print("Accuracy: %0.2f percent" % (acc *100))

(4962, 80072)
Accuracy: 92.52 percent


**Model accuracy is `92.52%`**

## 6. Include the Subject in the analysis of the emails, does the accuracy/performance of the model increase?

#### Create a new column containing the concatenation of email `Subject` and `Body`.

In [31]:
enron['Total'] = [sub + bod for (sub, bod) in zip(enron.iloc[:,0],enron.iloc[:,1])]

In [32]:
enron.head(5)

Subject  \
0  dobmeos with hgh my energy level has gone up !...   
1        your prescription is ready . . oxwq s f e\n   
2                            get that new car 8434\n   
3                              await your response\n   
4  coca cola , mbna america , nascar partner with...   

                                                Body  Spam  \
0  introducing\n doctor - formulated\n hgh\n huma...     1   
1  low cost prescription medications\n soma , ult...     1   
2  people nowthe weather or climate in any partic...     1   
3  dear partner ,\n we are a team of government o...     1   
4  stock\n profile\n about\n company\n investment...     1   

                                               Total  
0  dobmeos with hgh my energy level has gone up !...  
1  your prescription is ready . . oxwq s f e\nlow...  
2  get that new car 8434\npeople nowthe weather o...  
3  await your response\ndear partner ,\n we are a...  
4  coca cola , mbna america , nascar partner with...

In [33]:
train_X, test_X, train_y, test_y = train_test_split(enron["Total"], enron["Spam"], test_size=0.3, random_state=0)

In [34]:
vectorizer = TfidfVectorizer(stop_words='english')
tfidf_train_X = vectorizer.fit_transform(train_X)

In [35]:
index_value={i[1]:i[0] for i in vectorizer.vocabulary_.items()}

In [36]:
fully_indexed = []
for row in tfidf_train_X:
    fully_indexed.append({index_value[column]:value for (column,value) in zip(row.indices,row.data)})

In [37]:
classifier = GaussianNB()
classifier.fit(tfidf_train_X.toarray(), train_y)

GaussianNB()

In [38]:
tfidf_test_X = vectorizer.transform(test_X)
print(tfidf_test_X.shape)
scores = cross_val_score(classifier, tfidf_test_X.toarray(), test_y, cv=5)
acc = scores.mean()
print("Accuracy: %0.2f percent" % (acc *100))

(4962, 81140)
Accuracy: 93.31 percent


**The accuracy improved by `0.79%`**